Added computeDryBaseline implementation in RenderAligningShiftedBox

lewinpauli opened the pull request:
<!--
Thanks for filing a pull request!
Reviewers are typically assigned within a week of filing a request.
To learn more about code review, see our documentation on Tree Hygiene: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
-->

Implements computeDryBaseline method in RenderAligningShiftedBox to fix layout failures when DropdownButtonFormField is used inside Wrap inside AlertDialog. The method calculates baseline using dry layout methods and accounts for child positioning within the parent's coordinate space.

Issue appeared when DropdownButtonFormField was used inside a Wrap inside Alertdialog

Fixes [flutter/flutter/issues/169214](https://github.com/flutter/flutter/issues/169214)


## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

lewinpauli committed:
Fix computeDryBaseline implementation in RenderShiftedBox

Implements computeDryBaseline method in RenderShiftedBox to fix layout
failures when DropdownButtonFormField is used inside Wrap inside AlertDialog.
The method calculates baseline using dry layout methods and accounts for
child positioning within the parent's coordinate space.

Fixes flutter/flutter/issues/169214

lewinpauli committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline

lewinpauli committed:
Added packages/flutter/test/rendering/shifted_box_test.dart for computeDryBaseline in RenderShiftedBox

lewinpauli committed:
added dev/integration_tests/alert_dialog_dropdown_crash/integration_test/app_test.dart to check if issue 169214 still occurs

lewinpauli committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline

lewinpauli committed:
moved computeDryBaseline to RenderAligningShiftedBox to remove not needed switch case

lewinpauli committed:
update tests to use RenderAligningShiftedBox instead of RenderShiftedBox

lewinpauli committed:
fix: update print statement in integration test to reference RenderAligningShiftedBox for issue 169214

lewinpauli committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline

lewinpauli committed:
added # PUBSPEC CHECKSUM: at the end of pubspec.yaml

lewinpauli committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline

justinmc reviewed (commented):
Maybe you don't need an integration test at all?

I'm still curious about the checksum though. I actually don't know how those are generated off the top of my head. I'll look for an answer either way.

*Edit*: I believe you need to run `flutter update-packages --update-hashes` to update that checksum. Thanks to @jason-simmons for helping me out with that.

lewinpauli committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline

lewinpauli committed:
just testing if "flutter update-packages --update-hashes" works to create # PUBSPEC CHECKSUM: (yes it does)

lewinpauli committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline

lewinpauli committed:
Removed integration test files, added widget test packages/flutter/test/material/dropdown_dialog_baseline_test.dart and update computeDryBaseline in RenderAligningShiftedBox to prevent crashes with DropdownButton in AlertDialog

lewinpauli commented:
Now a unittest: packages/flutter/test/rendering/shifted_box_test.dart

and a widget test extist: packages/flutter/test/material/dropdown_dialog_baseline_test.dart

to verify computeDryBaseline in RenderAlignedShiftedBox exists, and the widget test reproduces the issue in https://github.com/flutter/flutter/issues/169214

lewinpauli committed:
moved test out of material directory and into rendering, also renamed to aligning_shifted_box_baseline_test.dart

lewinpauli committed:
Removed redundant alignment parameter from RenderPositionedBox in computeDryBaseline tests

lewinpauli committed:
Refactor RenderAligningShiftedBox to use getDryLayout instead of computeDryLayout and remove unused aligning_shifted_box_baseline_test.dart in material directory

lewinpauli committed:
Remove unused empty dropdown_dialog_baseline_test.dart file from material directory

lewinpauli committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline

justinmc reviewed (commented):
Thanks for moving the integration tests! It looks like it's thoroughly tested now.

Check out the failures, it looks like there is a legitimate problem with how the dry baseline is calculated.

lewinpauli committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline

lewinpauli committed:
using computeDryLayout instead of getDryLayout for self-sizing, and added . at the end of comments

lewinpauli commented on packages/flutter/lib/src/rendering/shifted_box.dart:
```diff
@@ -363,6 +363,28 @@ abstract class RenderAligningShiftedBox extends RenderShiftedBox {
+    if (child == null) {
+      return null;
+    }
+    // Use loosened constraints for the child, just like in performLayout
```
Just added that

lewinpauli committed:
Implement computeDryBaseline in RenderFittedBox and RenderAligningShiftedBox, and refactor size calculations in overflow boxes.

lewinpauli committed:
Add tests for computeDryBaseline in RenderFittedBox and RenderPositionedBox, covering various scenarios and edge cases

lewinpauli committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline

lewinpauli committed:
Refactor RenderPositionedBox layout calculations and update test cases for computeDryBaseline with improved BoxConstraints

lewinpauli committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline

lewinpauli committed:
removed computeDryBaseline of RenderFittedBox because its not part of this PR and removed test that I created

lewinpauli committed:
Add computeDryBaseline method to RenderFittedBox for baseline retrieval, to resolve testing errors

lewinpauli committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline

lewinpauli committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline

lewinpauli committed:
changed the computeDryBaseline implementations back to use getDryLayout(constraints) instead of manually calculating

lewinpauli committed:
Refactor computeDryBaseline and layout logic in RenderShiftedBox for consistency also fixed computeDistanceToActualBaseline to make flutter test packages/flutter/test/material/toggle_buttons_test.dart pass

lewinpauli committed:
Refactor size computation in RenderShiftedBox and add computeDryBaseline to RenderAnimatedSize for improved layout handling, also _RenderInputPadding _computeSize switched up height & width

lewinpauli committed:
Replace computeDryLayout with getDryLayout in RenderAnimatedSize for accurate size calculations

lewinpauli committed:
trying to pass test/material/time_picker_test.dart

lewinpauli committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline

lewinpauli committed:
Fix dry/actual baseline mismatches; top-align baseline-less Flex

rendering/flex:

Add dry baseline simulation helpers mirroring performLayout.
In horizontal baseline mode, align children without baselines to the top (ignore verticalDirection) and mirror in dry path.
rendering/shifted_box:

RenderPadding.computeDryBaseline: return child baseline + padding.top; null if child has none.
RenderSizedOverflowBox.computeDistanceToActualBaseline: include child offset.dy when present.
Tidy base comment in RenderShiftedBox.computeDryBaseline.
rendering/proxy_box:

computeDryBaseline: use child’s dry baseline when available, otherwise fall back to super.
material/badge:

_RenderBadge: implement computeDryLayout and computeDryBaseline to mirror performLayout (alignment, explicit offset, label vertical shift).
material/time_picker:

_RenderInputPadding.computeDryBaseline: compute center offset using dry size and add to child baseline.
semantics/binding:

Safely invoke platformDispatcher.setSemanticsTreeEnabled in tests via dynamic call, ignore when unavailable.
Rationale:

Make computeDryBaseline match actual layout offsets across wrappers.
Fix baseline alignment invariants in Flex and wrappers.
Eliminate “computeDistanceToActualBaseline differs from computeDryBaseline” assertions.
Tests:

rendering/flex “children with no baselines are top-aligned” passes.
material badge/time_picker baseline assertions resolved.
No functional changes to Cupertino filter behavior.

currently only flutter test packages/flutter/test/cupertino/popup_surface_test.dart -p vm --plain-name 'Saturation is applied before blur' fails

lewinpauli committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline

lewinpauli committed:
fixing analyze issues

lewinpauli committed:
fixing analyze issue: Refactor RenderAnimatedSize to use getDryLayout for size calculation instead of computeDryLayout

lewinpauli committed:
fiy analyze issue: RenderAnimatedSize computeDryBaseline replace computeDryLayout with getDryLayout

lewinpauli committed:
Merge upstream Flutter master into fix-renderShiftedBox-computeDryBaseline

Resolved merge conflicts in packages/flutter/lib/src/semantics/binding.dart

Removed semantics_binding_set_semantics_tree_enabled_test.dart (deleted upstream)

lewinpauli committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline

lewinpauli commented:
To provide some context, I originally wanted to fix an issue that appeared when a DropDownField was used inside a Wrap inside an AlertDialog, as described in [https://github.com/flutter/flutter/issues/169214](https://github.com/flutter/flutter/issues/169214). The issue was that the implementation of the 'computeDryBaseline' function was missing in the 'RenderAligningShiftedBox' class. I added this, as well as unit and widget tests.

Quite a few tests were failing, so I had to edit some files. For example, the renderAnimatedSize also had no computeDryBaseline implementation, as did RenderShiftedBox, RenderPositionedBox and _RenderBadge.

In material/button.dart, material/button_style_button.dart and _RenderInputPadding, the size (width, height) was mixed up.

lewinpauli committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline

lewinpauli committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline

lewinpauli committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline

justinmc reviewed (approved):
LGTM as long as Google tests pass. Thank you for all of the work and iterations on this. The lack of (or incorrect) dry layout calculations were probably causing headaches for a lot of users.

chunhtai reviewed (commented):
LGTM once justin's comments are addressed

dkwingsmt committed:
Merge branch 'master' into fix-renderShiftedBox-computeDryBaseline
